File: src/main/java/pl/kamilprzenioslo/safety/Examples.java
package pl.kamilprzenioslo.safety;

import pl.kamilprzenioslo.k.Contact;
import pl.kamilprzenioslo.k.Employee;
import pl.kamilprzenioslo.k.Task;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;
import java.net.UnknownHostException;
import java.sql.SQLException;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

import static java.util.Comparator.naturalOrder;
import static pl.kamilprzenioslo.k.SafeCallKt.exampleTask;
import static pl.kamilprzenioslo.k.safety.ExamplesKt.throwsExceptionCheckedInJava;

public class Examples {

    void printStringLength(String string) {
        int stringLength = 0;

        if (string != null) {
            stringLength = string.length();
        }

        System.out.println(stringLength);
    }

    void conditionals() {
        var assignee =
                exampleTask().getAssignee();
        if (assignee != null) {
            var contact =
                    assignee.getContact();
            if (contact != null) {
                contact.getEmail();
            }
        }
    }

    void optionalTask() {
        Optional.of(exampleTask())
                .map(Task::getAssignee)
                .map(Employee::getContact)
                .map(Contact::getEmail)
                .orElse(null);
    }

    int stringLength(String string) {
        return string != null ? string.length() : 0;
        // lub
        if (string != null) {
            return string.length();
        } else {
            throw new IllegalArgumentException();
        }
    }

    void newSocket() throws IOException {
        new Socket((String) null, 8080, InetAddress.getLocalHost(), 0);
    }

    // metoda zwracająca Optional
    Optional<String> findLongestString(Collection<String> strings) {
        return strings.stream().max(naturalOrder());
    }

    void optionalUsage() {
        // wykorzystanie otrzymanej instancji Optional
        String longestString = findLongestString(someStrings())
                .orElse("No longest string - collection was empty");
        // lub
        String longestString = findLongestString(someStrings()).orElseThrow();
        // lub
        findLongestString(someStrings()).ifPresent(System.out::println);
    }

    List<String> someStrings() {
        return List.of("a", "b");
    }

    void errorHandling() {
        try (FileReader reader = new FileReader("filename")) {
            // wykonywany kod
        } catch (FileNotFoundException ex) {
            // obsługa rzuconego wyjątku
        } catch (IOException | SQLException ex) {
            // obsługa rzuconego wyjątku
        } catch (Exception ex) {
            // obsługa rzuconego wyjątku
        } finally {
            // opcjonalny blok finalizujący
        }
    }

    void throwsExceptionChecked() {
        try {
            throwsExceptionCheckedInJava();
        } catch (IOException ex) {
            // obsługa rzuconego wyjątku
        }
    }
}

File: src/main/kotlin/pl/kamilprzenioslo/k/safety/Examples.kt
package pl.kamilprzenioslo.k.safety

import java.io.FileNotFoundException
import java.io.FileReader
import java.io.IOException
import java.sql.SQLException

fun nullAssignment() {
    var a: String = "abc"
    a = null // błąd kompilacji

    var b: String? = "abc"
    b = null // udana kompilacja
}

fun reassignment() {
    val a: Int? = 123
    val b: Int = a // błąd kompilacji

    val x: Int = 123
    val y: Int? = x // udana kompilacja
}

fun doublebang() {
    val string: String? = functionReturningANullableString()

    val stringLength = string.length // błąd kompilacji

    val forcedStringLength = string!!.length // udana kompilacja
}

fun functionReturningANullableString(): String? = null

fun stringLength(string: String?): Int {
    return string?.length ?: 0
    // lub
    return string?.length ?: throw IllegalArgumentException()
}

// metoda zwracająca wartość potencjalnie pustą
fun findLongestString(strings: Collection<String>): String? {
       return strings.maxOrNull();
   }

fun safeCallUsage() {
    val strings = listOf("a", "b")

    // wykorzystanie otrzymanej wartości potencjalnie pustej
    val longestString = findLongestString(strings) ?:
            "No longest string - collection was empty"
    // lub
    val longestString = findLongestString(strings) ?:
            throw NoSuchElementException()
    // lub
    findLongestString(strings)?.let { println(it) }
}

fun errorHandling() {
    try {
        FileReader("filename").use { reader ->
            // wykonywany kod
        }
    } catch (ex: FileNotFoundException) {
        // obsługa rzuconego wyjątku
    } catch (ex: IOException) {
        // obsługa rzuconego wyjątku
    } catch (ex: SQLException) {
        // obsługa rzuconego wyjątku
    } catch (ex: Exception) {
        // obsługa rzuconego wyjątku
    } finally {
        // opcjonalny blok finalizujący
    }
}

@Throws(IOException::class)
fun throwsExceptionCheckedInJava() {
    throw IOException()
}
